question2/odbcdemo.c

```diff
@@ -0,0 +1,288 @@
+    if (fgets(buffer, buflen, stdin) != 0) {
+        buffer[strcspn(buffer, "\n")] = '\0';
+        return buffer;
+    }
+    return 0;
+}
+
+/*
+** Defines used by main program.
+*/
+#define PWD_LEN 32
+#define UID_LEN 32
+#define DSN_LEN 32
+#define USAGE_MSG1 "Usage: %s \n"
+
+/*
+** Program: odbc demo
+**
+** Purpose: ODBC Demo routine.
+*/
+
+typedef struct {
+    UCHAR charCol1[1024];
+    // SQLLEN length1 ;
+    long length1;
+    short length2;
+} DataInfoStruct;
+
+DataInfoStruct dataStruct[32];
+DataInfoStruct dataStruct2[32];
+DataInfoStruct dataStruct3[32];
+
+int main(int argc, char *argv[]) {
+    SQLHANDLE hdbc;
+    SQLHANDLE henv;
+    SQLHANDLE hstmt;
+    RETCODE rc;
+    UCHAR uid[UID_LEN];
+    UCHAR pwd[PWD_LEN];
+    UCHAR driver[DSN_LEN];
+    UCHAR ver[32];
+    SQLSMALLINT strLen;
+    int i;
+
+    uid[0] = 0;
+    pwd[0] = 0;
+    if (argc > 1) {
+        printf(USAGE_MSG1, argv[0]);
+        return (1);
+    }
+
+    printf("\nEnter the DSN : ");
+    fgets_wrapper((char *)driver, DSN_LEN);
+    printf("\nEnter the UID : ");
+    fgets_wrapper((char *)uid, UID_LEN);
+    printf("\nEnter the PWD : ");
+    fgets_wrapper((char *)pwd, PWD_LEN);
+
+    printf("%s: will connect to data source '%s' as user '%s/%s'.\n", argv[0],
+           driver, uid, pwd);
+
+    rc = SQLAllocHandle(SQL_HANDLE_ENV, SQL_NULL_HANDLE, &henv);
+    if ((rc != SQL_SUCCESS) && (rc != SQL_SUCCESS_WITH_INFO)) {
+        printf("Unable to allocate environment\n");
+        exit(255);
+    }
+
+    rc = SQLSetEnvAttr(henv, SQL_ATTR_ODBC_VERSION, (SQLPOINTER)SQL_OV_ODBC3,
+                       SQL_IS_INTEGER);
+
+    rc = SQLConnect(hdbc, (SQLCHAR *)uid, SQL_NTS, (SQLCHAR *)pwd, SQL_NTS);
+    if ((rc != SQL_SUCCESS) && (rc != SQL_SUCCESS_WITH_INFO)) {
+        printf("SQLConnect: Failed...\n");
+        ODBC_error(henv, hdbc, SQL_NULL_HSTMT);
+        exit(255); /* Exit with failure */
+    }
+
+    rc = SQLAllocHandle(SQL_HANDLE_STMT, hdbc, &hstmt);
+    if ((rc != SQL_SUCCESS) && (rc != SQL_SUCCESS_WITH_INFO)) {
+        printf("Unable to Allocate a SQLHANDLE:\n");
+        ODBC_error(henv, hdbc, hstmt);
+        EnvClose(henv, hdbc);
+        exit(255);
+    }
+
+    rc =
+        SQLGetInfo(hdbc, SQL_DRIVER_VER, (SQLPOINTER)ver, sizeof(ver), &strLen);
+    if ((rc != SQL_SUCCESS) && (rc != SQL_SUCCESS_WITH_INFO)) {
+        printf("SQLGetInfo has Failed. RC=%d\n", rc);
+        ODBC_error(henv, hdbc, hstmt);
+    }
+
+    printf("Driver version: %s\n", ver);
+    printf("Calling SQLGetTypeInfo...\n");
+    rc = SQLGetTypeInfo((SQLHSTMT)hstmt, SQL_ALL_TYPES);
+    if ((rc != SQL_SUCCESS) && (rc != SQL_SUCCESS_WITH_INFO)) {
+        printf("SQLGetTypeInfo has Failed. RC=%d\n", rc);
+        ODBC_error(henv, hdbc, hstmt);
+        EnvClose(henv, hdbc);
+        exit(255);
+    }
+
+    rc = SQLBindCol(hstmt, 1, SQL_C_CHAR, &dataStruct[0].charCol1[0],
+                    (SDWORD)sizeof(dataStruct[0].charCol1),
```

Comment:
we should cast to the type expected in the API
ref: https://learn.microsoft.com/en-us/sql/odbc/reference/syntax/sqlbindcol-function?view=sql-server-ver16
```suggestion
                    (SQLLEN)sizeof(dataStruct[0].charCol1),
```